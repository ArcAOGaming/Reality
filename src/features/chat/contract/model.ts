import { ArweaveId } from "@/features/arweave/lib/model";
import { z } from "zod";

// Placeholder
// TODO: Define this properly
export const Message = z.object({
  MessageId: ArweaveId,
  Timestamp: z.number(),
  AuthorId: ArweaveId,
  AuthorName: z.string(),
  Content: z.string(),
});
export type Message = z.infer<typeof Message>;

export const MessageCreate = Message.omit({
  MessageId: true,
  Timestamp: true,
  AuthorId: true,
});
export type MessageCreate = z.infer<typeof MessageCreate>;
