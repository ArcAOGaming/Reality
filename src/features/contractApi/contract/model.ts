// Placeholder
// TODO: Define this properly
import { RJSFSchema } from '@rjsf/utils';
import { z } from 'zod';

export const ApiSchemaMethod = z.object({
  Title: z.string(),
  Description: z.string(),
  Schema: z.object({
    Tags: z.custom<RJSFSchema>(),
  }),
});
export type ApiSchemaMethod = z.infer<typeof ApiSchemaMethod>;

export const ApiSchema = z.record(ApiSchemaMethod);
export type ApiSchema = z.infer<typeof ApiSchema>;
