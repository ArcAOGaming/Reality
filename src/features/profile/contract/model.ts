import { ArweaveId } from "@/features/arweave/lib/model";
import { z } from "zod";

export const ProfileInfo = z.object({
  ProfileId: ArweaveId,
  Username: z.string(),
  ProfileImage: z.string(),
  CoverImage: z.string(),
  Description: z.string(),
  DisplayName: z.string(),
});
export type ProfileInfo = z.infer<typeof ProfileInfo>;

export const ProfileInfoCreate = ProfileInfo.omit({
  ProfileId: true,
  Username: true,
})
  .extend({
    UserName: z.string(),
    DateCreated: z.number(),
    DateUpdated: z.number(),
  })
  .partial({
    // Username: true,
    // ProfileImage: true,
    // CoverImage: true,
    // Description: true,
    // DisplayName: true,
  });
export type ProfileInfoCreate = z.infer<typeof ProfileInfoCreate>;

export const ProfileInfoUpdate = ProfileInfoCreate.omit({
  DateCreated: true,
}).partial();
export type ProfileInfoUpdate = z.infer<typeof ProfileInfoUpdate>;
