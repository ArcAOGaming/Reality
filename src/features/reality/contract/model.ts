import { z } from "zod";
import { _2dTileParams } from "./_2dTile";
import { ArweaveAddress } from "@/features/arweave/lib/model";

export const RealityVector = z.array(z.number());
export type RealityVector = z.infer<typeof RealityVector>;

export const RealityInfo = z.object({
  Parent: z.string(),
  Name: z.string(),
  Dimensions: z.number(),
  "Render-With": z.string(),
});
export type RealityInfo = z.infer<typeof RealityInfo>;

export const RealityParameterBounds = z.object({
  Lower: z.array(z.number()),
  Upper: z.array(z.number()),
});

export const RealityParameters = z.object({
  Bounds: z.optional(RealityParameterBounds),
  "2D-Tile-0": z.optional(_2dTileParams),
});
export type RealityParameters = z.infer<typeof RealityParameters>;

export const RealityEntityType = z.enum(["Unknown", "Avatar", "Hidden"]);
export type RealityEntityType = z.infer<typeof RealityEntityType>;

export const RealityEntityPosition = RealityVector;
export type RealityEntityPosition = z.infer<typeof RealityEntityPosition>;

export const DefaultInteraction = z.object({
  Type: z.literal("Default"),
});
export type DefaultInteraction = z.infer<typeof DefaultInteraction>;

export const Warp = z.object({
  Type: z.literal("Warp"),
  Size: z.optional(z.array(z.number())),
});
export type Warp = z.infer<typeof Warp>;

export const SchemaForm = z.object({
  Type: z.literal("SchemaForm"),
  Id: z.string(),
});
export type SchemaForm = z.infer<typeof SchemaForm>;

export const SchemaExternalForm = z.object({
  Type: z.literal("SchemaExternalForm"),
  Id: z.string(),
});
export type SchemaExternalForm = z.infer<typeof SchemaForm>;

export const RealityEntityMetadata = z.intersection(
  z.object({
    Interaction: z.optional(
      z.discriminatedUnion("Type", [
        DefaultInteraction,
        Warp,
        SchemaForm,
        SchemaExternalForm,
      ]),
    ),
    ProfileId: z.optional(z.string()),
    DisplayName: z.optional(z.string()),
    SkinNumber: z.optional(z.number()),
    SpriteTxId: z.optional(z.string()),
  }),
  z.record(z.string(), z.any()),
);
export type RealityEntityMetadata = z.infer<typeof RealityEntityMetadata>;

export const RealityEntity = z.object({
  Position: RealityEntityPosition,
  Type: RealityEntityType,
  Metadata: z.optional(RealityEntityMetadata),
  StateCode: z.optional(z.number()),
});
export type RealityEntity = z.infer<typeof RealityEntity>;

export const RealityEntityCreate = z.object({
  Type: z.optional(RealityEntityType),
  Position: z.optional(RealityEntityPosition),
  Metadata: z.optional(RealityEntityMetadata),
});
export type RealityEntityCreate = z.infer<typeof RealityEntityCreate>;

export const RealityEntities = z.record(ArweaveAddress, RealityEntity);
export type RealityEntities = z.infer<typeof RealityEntities>;

export const RealityEntityKeyed = RealityEntities.refine(
  (entities) => Object.keys(entities).length === 1,
  {
    message: "Expected exactly one entity",
  },
);
export type RealityEntityKeyed = z.infer<typeof RealityEntityKeyed>;
