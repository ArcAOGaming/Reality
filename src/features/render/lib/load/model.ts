import { ProfileRegistryClient } from "@/features/profile/contract/profileRegistryClient";
import { RealityClient } from "@/features/reality/contract/realityClient";

export type WorldState = {
  info: Awaited<ReturnType<RealityClient["readInfo"]>>;
  parameters: Awaited<ReturnType<RealityClient["readParameters"]>>;
  entities: Awaited<ReturnType<RealityClient["readEntitiesStatic"]>>;
  profiles: Awaited<ReturnType<ProfileRegistryClient["readProfiles"]>>;
};
