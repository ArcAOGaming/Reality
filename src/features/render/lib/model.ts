export type Size2D = {
  w: number;
  h: number;
};

export type Point2D = {
  x: number;
  y: number;
};

export type BoxCentered = {
  center: Point2D;
  edgeLength: number;
};
