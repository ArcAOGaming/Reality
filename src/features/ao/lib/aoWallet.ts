import { createDataItemSigner } from "@permaweb/aoconnect";

export type AoSigner = ReturnType<typeof createDataItemSigner>

export type AoWallet = {
  address: string;
  signer: AoSigner;
}

export type AoWalletConnectionResult = {
  success: true;
  result: AoWallet;
} | {
  success: false;
  error: unknown;
}

// TODO: Import these from somewhere more official
export type PermissionType =
  | "ACCESS_ADDRESS"
  | "ACCESS_PUBLIC_KEY"
  | "ACCESS_ALL_ADDRESSES"
  | "SIGN_TRANSACTION"
  | "ENCRYPT"
  | "DECRYPT"
  | "SIGNATURE"
  | "ACCESS_ARWEAVE_CONFIG"
  | "DISPATCH";

export interface AppInfo {
  name?: string;
  logo?: string;
}

export interface GatewayConfig {
  host: string;
  port: number;
  protocol: "http" | "https";
}

export type AoWalletConnector = (
  permissions: PermissionType[],
  appInfo?: AppInfo,
  gateway?: GatewayConfig,
) => Promise<AoWalletConnectionResult>;
