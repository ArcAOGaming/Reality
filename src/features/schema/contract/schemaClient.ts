import { connect } from "@/features/ao/lib/aoConnection";
import {
  AoContractClient,
  createAoContractClient,
} from "../../ao/lib/aoContractClient";
import { Schema, SchemaExternal } from "./model";
import { AoWallet } from "@/features/ao/lib/aoWallet";

export type SchemaClient = {
  aoContractClient: AoContractClient;

  // Reads
  readSchema(): Promise<Schema>;
  readSchemaExternal(): Promise<SchemaExternal>;
};

// Placeholder
// TODO: Define these methods properly
export const createSchemaClient = (
  aoContractClient: AoContractClient,
): SchemaClient => ({
  aoContractClient: aoContractClient,

  // Read
  readSchema: () =>
    aoContractClient.messageDelayReplyOneJson<Schema>(
      {
        tags: [{ name: "Action", value: "Schema" }],
      } /* Schema */,
    ),
  readSchemaExternal: () =>
    aoContractClient.messageDelayReplyOneJson<SchemaExternal>(
      {
        tags: [{ name: "Action", value: "SchemaExternal" }],
      } /* SchemaExternal */,
    ),
});

export const createSchemaClientForProcess =
  (wallet: AoWallet) => (processId: string) => {
    const aoContractClient = createAoContractClient(
      processId,
      connect(),
      wallet,
    );
    return createSchemaClient(aoContractClient);
  };
