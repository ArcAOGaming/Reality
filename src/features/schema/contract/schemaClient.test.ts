import { describe, test, expect, beforeAll } from "vitest";
import { createSchemaClient } from "./schemaClient";
import {
  AoContractClient,
  createAoContractClient,
} from "@/features/ao/lib/aoContractClient";
import { loadTestWallet } from "@/features/ao/test/lib/fsWallet";
import { AoWallet } from "@/features/ao/lib/aoWallet";
import { connect } from "@permaweb/aoconnect";
import { Schema } from "./model";

describe("createSchemaClient", () => {
  let testWallet: AoWallet;
  let schemaTestAoContractClient: AoContractClient;

  beforeAll(async () => {
    testWallet = await loadTestWallet();
    schemaTestAoContractClient = createAoContractClient(
      import.meta.env.VITE_LLAMAASSISTANT_PROCESS_ID,
      connect(),
      testWallet,
    );
  });

  test("creates client", async () => {
    const client = createSchemaClient(schemaTestAoContractClient);
    expect(client).toMatchSnapshot();
  });

  test("read Schema", async () => {
    const client = createSchemaClient(schemaTestAoContractClient);

    const schema = await client.readSchema();
    expect(Schema.safeParse(schema).success).toBe(true);
    expect(schema).toMatchSnapshot();
  });
});
