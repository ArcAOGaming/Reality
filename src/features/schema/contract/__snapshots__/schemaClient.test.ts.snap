// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSchemaClient > creates client 1`] = `
{
  "aoContractClient": {
    "aoClient": {
      "assign": [Function],
      "dryrun": [Function],
      "message": [Function],
      "monitor": [Function],
      "result": [Function],
      "results": [Function],
      "spawn": [Function],
      "unmonitor": [Function],
    },
    "aoWallet": {
      "address": "HHz5nNVEzkALjmZq8cdIdubMBJCpIFOvzO5uHfU5nQI",
      "anonymous": false,
      "signer": [Function],
      "type": "Keyfile",
    },
    "dryrunReadReplyOne": [Function],
    "dryrunReadReplyOneJson": [Function],
    "dryrunReadReplyOptional": [Function],
    "message": [Function],
    "processId": "y_obnQk5mkphKlulM7v1Xyn6IhJKZGhP_BC1qLJq46w",
  },
  "readSchema": [Function],
  "readSchemaExternal": [Function],
}
`;
